From e94b4f23277f7572aacbbeae50b8927e03be148a Mon Sep 17 00:00:00 2001
From: Petr Vorel <pvorel@suse.cz>
Date: Thu, 15 Jul 2021 17:35:27 +0200
Subject: [PATCH] osdep: Introduce include/grub/osdep/major.h and use it
... to factor out fix for glibc 2.25 introduced in 7a5b301e3 (build: Use
AC_HEADER_MAJOR to find device macros).
Note: Once glibc 2.25 is old enough and this fix is not needed also
AC_HEADER_MAJOR in configure.ac should be removed.
Signed-off-by: Petr Vorel <pvorel@suse.cz>
Reviewed-by: Daniel Kiper <daniel.kiper@oracle.com>
---
 configure.ac                         |  2 +-
 grub-core/osdep/devmapper/getroot.c  |  7 +-----
 grub-core/osdep/devmapper/hostdisk.c |  7 +-----
 grub-core/osdep/linux/getroot.c      |  7 +-----
 grub-core/osdep/unix/getroot.c       |  7 +-----
 include/grub/osdep/major.h           | 33 ++++++++++++++++++++++++++++
 6 files changed, 38 insertions(+), 25 deletions(-)
 create mode 100644 include/grub/osdep/major.h
Index: grub2/configure.ac
===================================================================
--- grub2.orig/configure.ac	2022-11-02 11:08:38.952223645 +0800
+++ grub2/configure.ac	2022-11-02 11:08:38.952223645 +0800
@@ -432,7 +432,7 @@
 
 # glibc 2.25 still includes sys/sysmacros.h in sys/types.h but emits deprecation
 # warning which causes compilation failure later with -Werror. So use -Werror here
-# as well to force proper sys/sysmacros.h detection.
+# as well to force proper sys/sysmacros.h detection. Used in include/grub/osdep/major.h.
 SAVED_CFLAGS="$CFLAGS"
 CFLAGS="$HOST_CFLAGS -Werror"
 AC_HEADER_MAJOR
Index: grub2/grub-core/osdep/devmapper/getroot.c
===================================================================
--- grub2.orig/grub-core/osdep/devmapper/getroot.c	2022-11-02 11:08:38.952223645 +0800
+++ grub2/grub-core/osdep/devmapper/getroot.c	2022-11-02 11:08:38.952223645 +0800
@@ -41,12 +41,7 @@
 #include <limits.h>
 #endif
 
-#if defined(MAJOR_IN_MKDEV)
-#include <sys/mkdev.h>
-#elif defined(MAJOR_IN_SYSMACROS)
-#include <sys/sysmacros.h>
-#endif
-
+#include <grub/osdep/major.h>
 #include <libdevmapper.h>
 
 #include <grub/types.h>
Index: grub2/grub-core/osdep/devmapper/hostdisk.c
===================================================================
--- grub2.orig/grub-core/osdep/devmapper/hostdisk.c	2022-11-02 11:08:38.952223645 +0800
+++ grub2/grub-core/osdep/devmapper/hostdisk.c	2022-11-02 11:08:38.952223645 +0800
@@ -12,6 +12,7 @@
 #include <grub/misc.h>
 #include <grub/i18n.h>
 #include <grub/list.h>
+#include <grub/osdep/major.h>
 
 #include <stdio.h>
 #include <stdlib.h>
@@ -25,12 +26,6 @@
 #include <errno.h>
 #include <limits.h>
 
-#if defined(MAJOR_IN_MKDEV)
-#include <sys/mkdev.h>
-#elif defined(MAJOR_IN_SYSMACROS)
-#include <sys/sysmacros.h>
-#endif
-
 #ifdef HAVE_DEVICE_MAPPER
 # include <libdevmapper.h>
 
Index: grub2/grub-core/osdep/linux/getroot.c
===================================================================
--- grub2.orig/grub-core/osdep/linux/getroot.c	2022-11-02 11:08:38.952223645 +0800
+++ grub2/grub-core/osdep/linux/getroot.c	2022-11-02 11:08:38.952223645 +0800
@@ -36,12 +36,7 @@
 #include <limits.h>
 #endif
 
-#if defined(MAJOR_IN_MKDEV)
-#include <sys/mkdev.h>
-#elif defined(MAJOR_IN_SYSMACROS)
-#include <sys/sysmacros.h>
-#endif
-
+#include <grub/osdep/major.h>
 #include <grub/types.h>
 #include <sys/ioctl.h>         /* ioctl */
 #include <sys/mount.h>
Index: grub2/grub-core/osdep/unix/getroot.c
===================================================================
--- grub2.orig/grub-core/osdep/unix/getroot.c	2022-11-02 11:08:38.952223645 +0800
+++ grub2/grub-core/osdep/unix/getroot.c	2022-11-02 11:08:38.952223645 +0800
@@ -52,12 +52,7 @@
 #endif /* ! FLOPPY_MAJOR */
 #endif
 
-#include <sys/types.h>
-#if defined(MAJOR_IN_MKDEV)
-#include <sys/mkdev.h>
-#elif defined(MAJOR_IN_SYSMACROS)
-#include <sys/sysmacros.h>
-#endif
+#include <grub/osdep/major.h>
 
 #if defined(HAVE_LIBZFS) && defined(HAVE_LIBNVPAIR)
 # include <grub/util/libzfs.h>
Index: grub2/include/grub/osdep/major.h
===================================================================
--- /dev/null	1970-01-01 00:00:00.000000000 +0000
+++ grub2/include/grub/osdep/major.h	2022-11-02 11:08:38.952223645 +0800
@@ -0,0 +1,33 @@
+/*
+ *  GRUB  --  GRand Unified Bootloader
+ *  Copyright (C) 2021  Free Software Foundation, Inc.
+ *
+ *  GRUB is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  GRUB is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with GRUB.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ *  Fix for glibc 2.25 which is deprecating the namespace pollution of
+ *  sys/types.h injecting major(), minor(), and makedev() into the
+ *  compilation environment.
+ */
+
+#ifndef GRUB_OSDEP_MAJOR_H
+#define GRUB_OSDEP_MAJOR_H	1
+
+#include <sys/types.h>
+
+#ifdef MAJOR_IN_MKDEV
+# include <sys/mkdev.h>
+#elif defined (MAJOR_IN_SYSMACROS)
+# include <sys/sysmacros.h>
+#endif
+#endif /* GRUB_OSDEP_MAJOR_H */
